### DISCUSSION TITLE
running async code from interrupt?

### DISCUSSION THREAD

**Discussion by ThinkTransit** [micropython/micropython#15478] [ESP32] [answered]
Is it possible to run an async function from a pin interrupt?

For example, not sure why this doesn't work, any ideas?

```python
async def go_to_sleep():
    # Custom code that requires async
    machine.deepsleep(60*1000)

def sleep_callback(p):
    asyncio.create_task(go_to_sleep)
    # asyncio.run(go_to_sleep)

p0 = Pin(14, Pin.IN)
p0.irq(trigger=Pin.IRQ_FALLING, handler=sleep_callback)
```

```bash
Traceback (most recent call last):
  File "main.py", line 35, in sleep_callback
  File "asyncio/core.py", line 1, in create_task
TypeError: coroutine expected
```

**Comment by bixb922** [comment, marked as answer]
The correct way to create a task is ```asyncio.create_task(go_to_sleep() )```, with () 

However, it's not a good idea to launch a task from a ISR (interrupt service routine) . Please see this excellent tutorial https://github.com/peterhinch/micropython-async/blob/master/v3/docs/INTERRUPTS.md

One good way to do this is to leave an asyncio task running always awaiting a ThreadSafeFlag that is set by the ISR.

**Reply by ThinkTransit** [reply]
Thanks I missed the ()!
